pootle/apps/pootle_fs/finder.py

```diff
@@ -0,0 +1,182 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import fnmatch
+import os
+import re
+
+import scandir
+
+from django.utils.functional import cached_property
+from django.utils.lru_cache import lru_cache
+
+
+PATH_MAPPING = (
+    (".", "\."),
+    ("<language_code>", "(?P<language_code>[\w\-\.]*)"),
+    ("<filename>", "(?P<filename>[\w\-\.]*)"),
+    ("/<dir_path>/", "/<dir_path>"),
+    ("<dir_path>", "(?P<dir_path>[\w\/\-]*?)"))
+
+DEFAULT_EXTS = ("po", "pot")
+
+
+class TranslationFileFinder(object):
+
+    path_mapping = PATH_MAPPING
+
+    def __init__(self, translation_path, path_filters=None, ext=None):
+        self.translation_path = translation_path
+        self.ext = ext or DEFAULT_EXTS
+        self.path_filters = path_filters
+        TranslationPathValidator(translation_path).validate()
+
+    @cached_property
+    def regex(self):
+        return re.compile(self._parse_path())
+
+    @cached_property
+    def file_root(self):
+        file_root = self.translation_path.split("<")[0]
+        if not file_root.endswith("/"):
+            file_root = "/".join(file_root.split("/")[:-1])
+        return file_root.rstrip("/")
+
+    def match_filepath(self, file_path):
```

Comment:
As always it would be nice to have some docstrings that provide a brief description of what it does. Not nice to always have to read the code to understand what it does.